agent/resource.go

```diff
@@ -0,0 +1,228 @@
+	if !ok {
+		if err != nil {
+			err = fmt.Errorf("%w: %w", ErrUnmanaged, err)
+		} else {
+			err = fmt.Errorf("%w: %w", apierrors.NewForbidden(gvr.GroupResource(), name, ErrUnmanaged), ErrUnmanaged)
+		}
+	}
+	return res, err
+}
+
+// getAvailableResources retrieves a list of available resources for a given
+// combination of group and version specified in gvr. The resource in the gvr
+// must be empty, otherwise an error is returned.
+func (a *Agent) getAvailableResources(ctx context.Context, gvr schema.GroupVersionResource, namespace string) (*unstructured.UnstructuredList, error) {
+	var err error
+	var res *unstructured.UnstructuredList
+
+	// We do not want resources to be listed, only APIs
+	switch gvr.Resource {
+	case "apiresources":
+	case "apigroups":
+	case "":
+		break
+	default:
+		return nil, fmt.Errorf("not authorized to list resources: %s", gvr.String())
+	}
+
+	rif := a.kubeClient.DynamicClient.Resource(gvr)
+
+	if namespace != "" {
+		res, err = rif.Namespace(namespace).List(ctx, v1.ListOptions{})
+	} else {
+		res, err = rif.List(ctx, v1.ListOptions{})
+	}
+
+	return res, err
+}
+
+// isResourceManaged checks whether a given resource is considered to be
+// managed by an Argo CD application.
+func isResourceManaged(kube *kube.KubernetesClient, res *unstructured.Unstructured, maxRecurse int) (bool, error) {
+	if maxRecurse < 1 {
+		return false, fmt.Errorf("recursion limit reached")
+	}
+
+	// If the resource carries a tracking label or annotation, regardless of
+	// its value, we deem the resource to be managed by Argo CD. At this
+	// point in time, we do not care about the particular details of the
+	// managing app.
+	refs := res.GetOwnerReferences()
+	lbls := res.GetLabels()
+	annt := res.GetAnnotations()
+
+	// TODO: Read Argo CD configuration for the actual value of the tracking
+	// method and the label name.
+	_, lok := lbls["app.kubernetes.io/instance"]
+	_, aok := annt["argocd.argoproj.io/tracking-id"]
+	if lok || aok {
+		return true, nil
+	}
+
+	// Recursively check any owners referenced by the inspected resource.
+	for _, ref := range refs {
+		gv, err := schema.ParseGroupVersion(ref.APIVersion)
+		if err != nil {
+			return false, err
+		}
+
+		rif, err := kube.DiscoveryClient.ServerResourcesForGroupVersion(gv.String())
+		if err != nil {
+			return false, err
+		}
+
+		var gvr schema.GroupVersionResource
+		for _, apiResource := range rif.APIResources {
+			if ref.Kind == apiResource.Kind {
+				gvr = schema.GroupVersionResource{Group: apiResource.Group, Version: apiResource.Version, Resource: apiResource.Name}
+				break
+			}
+		}
+
+		// This should not happen, as it means we couldn't resolve the owner's
+		// Kind to a resource available in the cluster.
+		if gvr.Resource == "" {
+			continue
+		}
+
+		// Get the owning object from the cluster.
+		owner, err := kube.DynamicClient.Resource(gvr).Namespace(res.GetNamespace()).Get(context.Background(), ref.Name, v1.GetOptions{})
```

Comment:
Do we need to handle the case of a namespace-scoped resource having a cluster-scoped owner? Since we use the namespace of `res` I guess the client would be restricted to access resources only in that namespace.

Reply:
Ah, good catch. I missed the fact that namespace-scoped dependants may specify cluster-scoped parents. Will fix it.